### ISSUE TITLE
Dvorak QWERTY ⌘: <C-w> sends ,

### ISSUE THREAD

**Issue by hacker-DOM** [kovidgoyal/kitty#5518]
Wow, great application. Well done and than you.

**Background**
I am using a Dvorak keyboard layout. In particular, Mac's builtin "Dvorak-QWERTY ⌘". It is quite popular, as it allows you to use Qwerty keys for things like Cmd-C and Cmd-V. I've been successfully using it for two months and have never had any problems with it.

**Describe the bug**
On Dvorak, the "," (comma) key is mapped to "w" (fwiw, comma is mapped to "w", it is one of the only length-2 cycles :))

I would expect that `<C-w>` (ctrl+w) would send `<C-w>`, e.g. to change windows in Vim. But just sends on unmodified comma (",") 🤷🏽‍♂️

Running the show_key kitten and pressing C-w (ctrl+w), I get:

US Layout (works on this layout):

```
Ctrl+LEFT_CONTROL PRESS 
CSI 57442 ; 5 u

Ctrl+, PRESS 
CSI 44 ; 5 u

Ctrl+, RELEASE 
CSI 44 ; 5 : 3 u

LEFT_CONTROL RELEASE 
CSI 57442 ; 1 : 3 u
```

Basic Dvorak layout (also works on this one):

```
Ctrl+LEFT_CONTROL PRESS 
CSI 57442 ; 5 u

Ctrl+w PRESS 
CSI 119 :  : 44 ; 5 u
Alternate key: , 

Ctrl+w RELEASE 
CSI 119 :  : 44 ; 5 : 3 u
Alternate key: , 

LEFT_CONTROL RELEASE 
CSI 57442 ; 1 : 3 u
```

Dvorak-QWERTY ⌘ (layout I am using):

```
Ctrl+LEFT_CONTROL PRESS 
CSI 57442 ; 5 u

Ctrl+w PRESS ,
CSI 119 :  : 44 ; 5 ; 44 u
Alternate key: , 

Ctrl+w RELEASE 
CSI 119 :  : 44 ; 5 : 3 u
Alternate key: , 

LEFT_CONTROL RELEASE 
CSI 57442 ; 1 : 3 u
```

It seems the alternate (physical) key is fsr being used on Dvorak-QWERTY ⌘. For the program `showkey -a`, I get:

`,` for US and Dvorak-QWERTY ⌘,

`<CTL-W=ETB>` for (Basic) Dvorak

**To Reproduce**
1. Set a Dvorak keyboard layout.
2. Run Nvim `nvim --clean foo.txt`
3. `^w`

Expected behavior: The symbols "^W" appear in the status bar. Actual behavior: nothing happens.

**Environment details**

~~One sec~~

```
kitty 0.26.2 created by Kovid Goyal
Darwin DominiksMBP2019.docsis.vodafone.cz 21.5.0 Darwin Kernel Version 21.5.0: Tue Apr 26 21:08:22 PDT 2022; root:xnu-8020.121.3~4/RELEASE_X86_64 x86_64
ProductName:	macOS ProductVersion:	12.4 BuildVersion:	21F79
Frozen: True
Paths:
  kitty: /Applications/kitty.app/Contents/MacOS/kitty
  base dir: /Applications/kitty.app/Contents/Resources/kitty
  extensions dir: /Applications/kitty.app/Contents/Resources/Python/lib/kitty-extensions
  system shell: zsh
Loaded config files:
  /Users/dteiml/.config/kitty/kitty.conf

Config options different from defaults:
font_size 15.0
kitty_mod ctrl+cmd
Removed shortcuts:
	ctrl+cmd+, →  load_config_file
	ctrl+cmd+f →  toggle_fullscreen
Changed shortcuts:
	kitty_mod+, →  load_config_file
	kitty_mod+f →  toggle_fullscreen

Important environment variables seen by the kitty process:
	PATH                                /Applications/kitty.app/Contents/MacOS:/usr/bin:/bin:/usr/sbin:/sbin
	LANG                                en_US.UTF-8
	SHELL                               zsh
	USER                                dteiml
```

**Note**
FWIW, I get the expected behavior in Terminal.app, Tabby, Hyper, iTerm2

**Comment by kovidgoyal** [comment]
Post the output from running kitty with --debug-input

**Comment by kovidgoyal** [comment]
I dont need show_key output, I need the output from running 

kitty --debug-input

**Comment by hacker-DOM** [comment]
I understand this stuff is complex, so it's okay if there isn't a builtin solution - From the docs I got the impression you have access to a Mac, so if you e.g. find a solution with mappings, that would be great as well. I personally wasn't able to find a quick solution with mappings in Alfred (Workflows), will investigate more tomorrow

**Comment by hacker-DOM** [comment]
> I dont need show_key output, I need the output from running 
> 
> 
> 
> kitty --debug-input

Maybe I'm being a noob, but I get `Unknown option: --debug-info`.

The above output I got with option+cmd+, (called "Debug `kitty.conf`" in the Overview section of the docs).

I'll shamelessly assume that is what you're looking for, correct me if I'm wrong

**Comment by kovidgoyal** [comment]
--debug-input not --debug-info

**Comment by hacker-DOM** [comment]
> --debug-input not --debug-info

(ctrl-w followed by cmd-v)

```
updateIMEPosition: left=2.000000, top=5.500000, width=9.500000, height=17.500000
updateIMEPosition: left=2.000000, top=75.500000, width=9.500000, height=17.500000
updateIMEPosition: left=2.000000, top=75.500000, width=9.500000, height=17.500000
Move x: 473.1 y: 229.1 grabbed: 0
-------------- flags changed -----------------
NSEvent: type=FlagsChanged loc=(236.553,369.439) time=38922.1 flags=0x40101 win=0x7fdda889f480 winNum=28906 ctxt=0x0 keyCode=59
flagsChanged: modifier: ctrl native_key: 0x3b (<noname>) glfw_key: 0xe062 mods: ctrl 
on_key_input: glfw key: 0xe062 native_code: 0x3b action: PRESS mods: ctrl text: '' state: 0 updateIMEPosition: left=21.000000, top=128.000000, width=9.500000, height=17.500000
ignoring as keyboard mode does not support encoding this event
---------------- key down -------------------
NSEvent: type=KeyDown loc=(236.553,369.439) time=38922.1 flags=0x40101 win=0x7fdda889f480 winNum=28906 ctxt=0x0 chars="" unmodchars="w" repeat=0 keyCode=43
Press: native_key: 0x2b (w) glfw_key: 0x77 mods: ctrl char_count: 1 deadKeyState: 0 repeat: 0 
	insertText: , replacementRange: (9223372036854775807, 0)
text: 0x2c glfw_key: w marked_text: ()
on_key_input: glfw key: 0x77 native_code: 0x2b action: PRESS mods: ctrl text: ',' state: 0 updateIMEPosition: left=21.000000, top=128.000000, width=9.500000, height=17.500000
sent key as text to child
----------------- key up --------------------
NSEvent: type=KeyUp loc=(236.553,369.439) time=38922.2 flags=0x40101 win=0x7fdda889f480 winNum=28906 ctxt=0x0 chars="" unmodchars="w" repeat=0 keyCode=43
Release: native_key: 0x2b (w) glfw_key: 0x77 mods: ctrl 
on_key_input: glfw key: 0x77 native_code: 0x2b action: RELEASE mods: ctrl text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
ignoring as keyboard mode does not support encoding this event
-------------- flags changed -----------------
NSEvent: type=FlagsChanged loc=(236.553,369.439) time=38922.2 flags=0x100 win=0x7fdda889f480 winNum=28906 ctxt=0x0 keyCode=59
flagsChanged: modifier: ctrl native_key: 0x3b (<noname>) glfw_key: 0xe062 mods: none 
on_key_input: glfw key: 0xe062 native_code: 0x3b action: RELEASE mods: none text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
ignoring as keyboard mode does not support encoding this event
-------------- flags changed -----------------
NSEvent: type=FlagsChanged loc=(236.553,369.439) time=38937.6 flags=0x100108 win=0x7fdda889f480 winNum=28906 ctxt=0x0 keyCode=55
flagsChanged: modifier: super native_key: 0x37 (<noname>) glfw_key: 0xe064 mods: super 
on_key_input: glfw key: 0xe064 native_code: 0x37 action: PRESS mods: super text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
ignoring as keyboard mode does not support encoding this event
---------------- key down -------------------
NSEvent: type=KeyDown loc=(236.553,369.439) time=38937.6 flags=0x100108 win=0x7fdda889f480 winNum=28906 ctxt=0x0 chars="v" unmodchars="k" repeat=0 keyCode=9
Press: native_key: 0x9 (k) glfw_key: 0x6b mods: super char_count: 1 deadKeyState: 0 repeat: 0 
	TextInputCtx: doCommandBySelector: (noop:)
text: <none> glfw_key: k marked_text: ()
on_key_input: glfw key: 0x6b native_code: 0x9 action: PRESS mods: super text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
handled as shortcut
----------------- key up --------------------
NSEvent: type=KeyUp loc=(236.553,369.439) time=38937.6 flags=0x100108 win=0x7fdda889f480 winNum=28906 ctxt=0x0 chars="v" unmodchars="k" repeat=0 keyCode=9
Release: native_key: 0x9 (k) glfw_key: 0x6b mods: super 
on_key_input: glfw key: 0x6b native_code: 0x9 action: RELEASE mods: super text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
ignoring release event for previous press that was handled as shortcut
-------------- flags changed -----------------
NSEvent: type=FlagsChanged loc=(236.553,369.439) time=38937.6 flags=0x100 win=0x7fdda889f480 winNum=28906 ctxt=0x0 keyCode=55
flagsChanged: modifier: super native_key: 0x37 (<noname>) glfw_key: 0xe064 mods: none 
on_key_input: glfw key: 0xe064 native_code: 0x37 action: RELEASE mods: none text: '' state: 0 updateIMEPosition: left=30.500000, top=128.000000, width=9.500000, height=17.500000
ignoring as keyboard mode does not support encoding this event
```

**Comment by kovidgoyal** [comment]
This is a bug in this layout's implementation. This layout is supposed
to switch back to qwerty only when cmd is held down.

ctrl+cmd+w triggers a global menubar shortcut, which is why it has no
effect in kitty.

ctrl+w is correctly handled as ctrl+,

Trying this layout in Apple's Terminal.app by running

kitty +kitten show_key

in Terminal.app

ctrl+w is incorrectly reported as ^W (aka ctrl+w) it should be reported
as just , as cmd is not held down.

cmd+ctrl+w does nothing in Terminal.app either, other than trigerring a
system beep.

I suggest if you want to use dvorak go the whole hog and dont try this 
hacky half way house.

Fundamentally, the entire concept of this layout is broken. It changes
key meaning on holding down cmd. That means if you want to press ctrl+w
using the w key instead of the comma key, you have to press cmd+ctrl+w.
The application then gets the key event from the system,
as cmd+ctrl+w not ctrl+w. There is no way for the application to know
that the cmd in this event is to be ignored.

**Comment by hacker-DOM** [comment]
When I was referring to keyboard shortcuts, I forgot to mention that I was using a Dvorak physical layout for the Ctrl+w (key in the bottom right quadrant), but a Qwerty physical layout for the Cmd+v examples (so J on my keyboard since I use a physical Dvorak layout). From you post I think you might have tested a different key bc of this lapse on my part

I'll take a deeper look at it over the weekend. It's not the most fun topic but I want it to work bc I like the app thus far

**Comment by hacker-DOM** [comment]
A little Appundix on why I don't want to switch Full Dvorak:

Even before I adopted Dvorak and was using Qwerty, I had the following realization:

Keys on the right side of the keyboard are rarely used in Cmd+ chords. You have Cmd+c, Cmd+v, Cmd+f... The only two on the right side of the keyboard that are sometimes used are Cmd+p (print) and Cmd+k (Command Pallete in some apps) and I figured I could live without those.

So I remmaped the right side of the keyboard to help touch typing.

E.g., Cmd+j goes down, Cmd+k goes up, Cmd+i is Carriage return, system-wide (in MacOS you can do this with Alfred or Hammerspoon).

Fast-forward to two months ago when I started using Dvorak.

If I remmaped the new keys - e.g., Cmd+t in place of Cmd+k, I would no longer be able to open new tabs. 😀 Generally, I wouldn't be able to use the observation that the right side is less important, because many of the left-side keys are now on the right...

I was about to abandon Dvorak when I discovered Dvorak Qwerty Cmd. It works fantastic! This is the first time I have a problem, and I should say again that all other terminals work as expected

If we can work together to fix this issue, I'll be able to keep Dvorak and system-wide shortcuts!

**Comment by kovidgoyal** [comment]
On Thu, Sep 22, 2022 at 08:59:43AM -0700, Dom wrote:
> I was about to abandon Dvorak when I discovered Dvorak Qwerty Cmd. It works fantastic! This is the first time I have a problem, and I should say again that all other terminals work as expected

I posted an example of Terminal.app not working as expected. And I
explained why this entire concept is fatally flawed.

> 
> If we can work together to fix this issue, I'll be able to keep Dvorak and system-wide shortcuts!

You are most welcome to try to fix it, if you can come up with a
reasonable PR I will be happy to review. But, as I explained before
IMO that is impossible given the fundamental problems with this layout.

**Comment by hacker-DOM** [comment]
Ok!

First, let me apologize again for not being 100% clear about the physical keys being pressed. In this post, "physical key" will be the QWERTY equivalent. E.g., if one's trying the examples on the Mac builtin keyboard, it will be the letter that is printed on the keyboard. "Interpreted key" will be the key presumably sent from the operating system to the application. E.g.:

PK (physical key): "v"
IK: "k" (on Dvorak)

**Background**

I wasn't able to reach the same conclusion, that it is a problem in the layout. I'll be quoting your text here, don't take that in a bad way; I just want to get to the bottom of this for myself, and maybe others. Some questions will be purely educational for me, feel free to ignore them - but I think they might be useful for others as well.

> This is a bug in this layout's implementation. This layout is supposed
to switch back to qwerty only when cmd is held down.

I don't think the layout has a bug. Let's forget the terminal for a sec. I opened a Chromium browser, navigated to "chrome://extensions/shortcuts", and tried assigning various shortcuts. I all works as expected. E.g.:

Physical keys: Cmd +s (Interpreted keys Cmd+s)

Physical keys: Alt +s (Interpreted keys Alt+o)

Physical keys: Ctrl+s (Interpreted keys Ctrl+o)

It works exactly as "advertized."

> ctrl+w is correctly handled as ctrl+,

I am not getting this behavior, at least not in the aforementioned browser. "ctrl-w" correctly becomes "ctrl-,", "ctrl-," correctly becomes "ctrl-w".

> Trying this layout in Apple's Terminal.app by running

> kitty +kitten show_key

> in Terminal.app

> ctrl+w is incorrectly reported as ^W (aka ctrl+w) 

I am getting the same behavior here. Both "ctrl-w" and "ctrl-," give the following output. This is perhaps weird behavior of Terminal.app, but I don't care about Terminal.app :-)

```
UNIX		send_text
^W		\x17
^W		\x17
```

> [ctrl-w] should be reported as just , as cmd is not held down.

Shouldn't it be reported as "ctrl-,"? Excuse my ignorance.

> Fundamentally, the entire concept of this layout is broken. It changes
> key meaning on holding down cmd. That means if you want to press ctrl+w
> using the w key instead of the comma key, you have to press cmd+ctrl+w.
> The application then gets the key event from the system,
> as cmd+ctrl+w not ctrl+w. There is no way for the application to know
> that the cmd in this event is to be ignored.

This is what made me realize that I was previosly erroneusly talking about interpreted keys.

In the context of *physical* keys, I am talking about "ctrl-,".

Hopefully we can agree that this should send "ctrl-w", i.e., window switcher in (Neo)Vim. The problem is it does not. :-(

I hate to beat the nail, but I do believe it is the strongest argument that there is unexpected behavior in our terminal: all applications I have tried, recieve "ctrl-w" on pressing "ctrl-,". Moreover in all terminals I tried, following the "Steps to reproduce" in the first post (except that the third step is "^," (as physical keys)), correctly allows me to switch windows in Vim.

**Temporary workaround**

I found the following solves both of my issues:

```
map ctrl+w send_text all \x17
map cmd+v paste_from_clipboard
```

TLDR I think my posts to date were misleading because I was talking about interpreted keys, but it came across as physical keys. As a result of this mistake we were talking about different keys :-) I hope with this new post you might reconsider the opinion that it is not bug.

**Comment by hacker-DOM** [comment]
One sec, I made a mistake :)

**Comment by hacker-DOM** [comment]
Ok!

First, let me apologize again for not being 100% clear about the physical keys being pressed. In this post, "physical key" will be the QWERTY equivalent. E.g., if one's trying the examples on the Mac builtin keyboard, it will be the letter that is printed on the keyboard. "Interpreted key" will be the key presumably sent from the operating system to the application. E.g.:

PK (physical key): "v"
IK: "k" (on Dvorak)

**Background**

I wasn't able to reach the same conclusion, that it is a problem in the layout. I'll be quoting your text here, don't take that in a bad way; I just want to get to the bottom of this for myself, and maybe others. Some questions will be purely educational for me, feel free to ignore them - but I think they might be useful for others as well.

> This is a bug in this layout's implementation. This layout is supposed
to switch back to qwerty only when cmd is held down.

I don't think the layout has a bug. Let's forget the terminal for a sec. I opened a Chromium browser, navigated to "chrome://extensions/shortcuts", and tried assigning various shortcuts. I all works as expected. E.g.:

Physical keys: Cmd +s    Interpreted keys Cmd+s

Physical keys: Alt +s    Interpreted keys Alt+o

Physical keys: Ctrl+s    Interpreted keys Ctrl+o

It works exactly as "advertized."

> ctrl+w is correctly handled as ctrl+,

Weirdly enough, it seems to be that "ctrl-w" just becomes "," (as does "ctrl-,"). However, I'm nat really talking about "ctrl-w", but "ctrl-," (see below).

> Trying this layout in Apple's Terminal.app by running

> kitty +kitten show_key

> in Terminal.app

> ctrl+w is incorrectly reported as ^W (aka ctrl+w) 

I am getting the same behavior here. Both "ctrl-w" and "ctrl-," give the following output. This is perhaps weird behavior of Terminal.app, but I don't care about Terminal.app :-)

```
UNIX		send_text
^W		\x17
^W		\x17
```

> [ctrl-w] should be reported as just , as cmd is not held down.

Shouldn't it be reported as "ctrl-,"? Excuse my ignorance.

> Fundamentally, the entire concept of this layout is broken. It changes
> key meaning on holding down cmd. That means if you want to press ctrl+w
> using the w key instead of the comma key, you have to press cmd+ctrl+w.
> The application then gets the key event from the system,
> as cmd+ctrl+w not ctrl+w. There is no way for the application to know
> that the cmd in this event is to be ignored.

This is what made me realize that I was previosly erroneusly talking about interpreted keys.

In the context of *physical* keys, I am talking about "ctrl-,".

Hopefully we can agree that this should send "ctrl-w", i.e., window switcher in (Neo)Vim. The problem is it does not. :-(

I hate to beat the nail, but I do believe it is the strongest argument that there is unexpected behavior in our terminal: all applications I have tried, recieve "ctrl-w" on pressing "ctrl-,". Moreover in all terminals I tried, following the "Steps to reproduce" in the first post (except that the third step is "^," (as physical keys)), correctly allows me to switch windows in Vim.

**Temporary workaround**

I found the following solves both of my issues:

```
map ctrl+w send_text all \x17
map cmd+v paste_from_clipboard
```

TLDR I think my posts to date were misleading because I was talking about interpreted keys, but it came across as physical keys. As a result of this mistake we were talking about different keys :-) I hope with this new post you might reconsider the opinion that it is not bug.

**Comment by kovidgoyal** [comment]
On Fri, Sep 23, 2022 at 08:52:55AM -0700, Dom wrote:
> Ok!
> 
> First, let me apologize again for not being 100% clear about the physical keys being pressed. In this post, "physical key" will be the QWERTY equivalent. E.g., if one's trying the examples on the Mac builtin keyboard, it will be the letter that is printed on the keyboard. "Interpreted key" will be the key presumably sent from the operating system to the application. E.g.:
> 
> PK (physical key): "v"
> IK: "k" (on Dvorak)
> 
> **Background**
> 
> I wasn't able to reach the same conclusion, that it is a problem in the layout. I'll be quoting your text here, don't take that in a bad way; I just want to get to the bottom of this for myself, and maybe others. Some questions will be purely educational for me, feel free to ignore them - but I think they might be useful for others as well.
> 
> > This is a bug in this layout's implementation. This layout is supposed
> to switch back to qwerty only when cmd is held down.
> 
> I don't think the layout has a bug. Let's forget the terminal for a sec. I opened a Chromium browser, navigated to "chrome://extensions/shortcuts", and tried assigning various shortcuts. I all works as expected. E.g.:
> 
> Physical keys: Cmd +s    Interpreted keys Cmd+s
> 
> Physical keys: Alt +s    Interpreted keys Alt+o
> 
> Physical keys: Ctrl+s    Interpreted keys Ctrl+o
> 
> It works exactly as "advertized."

Try cmd+ctrl+whatever

> 
> > ctrl+w is correctly handled as ctrl+,
> 
> Weirdly enough, it seems to be that "ctrl-w" just becomes "," (as does "ctrl-,"). However, I'm nat really talking about "ctrl-w", but "ctrl-," (see below).

That is because traditional terminals cannot represent ctrl-, only kitty
can.

> 
> > Trying this layout in Apple's Terminal.app by running
> 
> > kitty +kitten show_key
> 
> > in Terminal.app
> 
> > ctrl+w is incorrectly reported as ^W (aka ctrl+w) 
> 
> I am getting the same behavior here. Both "ctrl-w" and "ctrl-," give the following output. This is perhaps weird behavior of Terminal.app, but I don't care about Terminal.app :-)
> 
> ```
> UNIX		send_text
> ^W		\x17
> ^W		\x17
> ```

The point is even Apple's software doesnt work with their own layout.

> 
> > [ctrl-w] should be reported as just , as cmd is not held down.
> 
> Shouldn't it be reported as "ctrl-,"? Excuse my ignorance.

No, because traditional terminals cannot report ctrl+,

> 
> > Fundamentally, the entire concept of this layout is broken. It changes
> > key meaning on holding down cmd. That means if you want to press ctrl+w
> > using the w key instead of the comma key, you have to press cmd+ctrl+w.
> > The application then gets the key event from the system,
> > as cmd+ctrl+w not ctrl+w. There is no way for the application to know
> > that the cmd in this event is to be ignored.
> 
> This is what made me realize that I was previosly erroneusly talking about interpreted keys.
> 
> In the context of *physical* keys, I am talking about "ctrl-,".
> 
> Hopefully we can agree that this should send "ctrl-w", i.e., window switcher in (Neo)Vim. The problem is it does not. :-(

This layout must send ctrl+dvorak key when you dont hold down cmd. And
when you do hold down command it must send cmd+ctrl+QWERTY key
And that's what it does in kitty but not in some other broken programs. 

As I explained before the whole concept behind this layout is totally
broken. I am not interested in supporting it. Since you found a hack to
satisfy yourself, feel free to continue using it. I am not spending
anymore time on this.

**Comment by hacker-DOM** [comment]
> Try cmd+ctrl+whatever

Ok this ending up convincing me! I acknowledge there's something weird with this layout XD

I ended up switching to full Dvorak & changing my system-wide keybindings to `Alt+`

**Comment by sidequestboy** [comment]
I was having trouble with this as well, so I made a karabiner configuration to remap command+* to qwerty using karabiner.ts: [here](https://github.com/sidequestboy/karabiner-config)
